apiserver/facades/client/application/application.go

```diff
@@ -451,26 +452,62 @@ func (api *APIBase) Deploy(args params.ApplicationsDeploy) (params.ErrorResults,
 }
 
 // compatibilityApplicationDeployArgs ensures that Deploy calls from
-// a juju 3.x client will work against a juju 2.9.x controller. In
-// juju 3.x, params.ApplicationsDeploy was changed to remove series
+// clients such as juju 3.x, python-libjuju 3.x, etc. work against a 2.9.x
+// controller. To support these clients, we make no assumptions about
+// which of the 3 potential places the OS is specified. We check all 3,
+// assert any provided match, and fill in the remain series.
+//
+// In juju 3.x, params.ApplicationsDeploy was changed to remove series
 // however, the facade version was not changed, nor was the name of
 // the params.ApplicationsDeploy changed. Thus it appears you can use
 // a juju 3.x client to deploy from a juju 2.9 controller, however it
-// fails because the series was not found. Make those corrections here.
+// fails because the series was not found.
+//
+// NOTE: We do not mess with charm url because this is dangerous. Mutating
+// charm url can lead to unexpected results
 func compatibilityApplicationDeployArgs(arg params.ApplicationDeploy) (params.ApplicationDeploy, error) {
-	origin := arg.CharmOrigin
-	if origin.Series != "" {
-		return arg, nil
+	s, err := getUnifiedSeries(arg)
+	if err != nil {
+		return arg, errors.Annotatef(err, "cannot deploy application")
 	}
-	originSeries, err := series.GetSeriesFromChannel(origin.Base.Name, origin.Base.Channel)
+	b, err := series.GetBaseFromSeries(s)
 	if err != nil {
-		return arg, err
+		return arg, errors.Trace(err)
 	}
-	origin.Series = originSeries
-	arg.Series = originSeries
+
+	arg.Series = s
+	arg.CharmOrigin.Series = s
+	arg.CharmOrigin.Base.Name = b.Name
+	arg.CharmOrigin.Base.Channel = b.Channel.String()
+
 	return arg, nil
 }
 
+// getUnifiedSeries checks the multiple places a series can be provided in a params.ApplicationDeploy,
+// checks they match if they're provided, and returns the matching series
+func getUnifiedSeries(arg params.ApplicationDeploy) (string, error) {
+	argSeries := arg.Series
+	originSeries := arg.CharmOrigin.Series
+	originBaseSeries, _ := series.GetSeriesFromChannel(arg.CharmOrigin.Base.Name, arg.CharmOrigin.Base.Channel)
+
+	s := argSeries
+	if s == "" {
+		s = originSeries
+	}
+	if s == "" {
+		s = originBaseSeries
+	}
+	if s == "" {
+		return "", errors.Errorf("no series provided")
```

Comment:
Is this gonna be a confusing error message for 3.x users where there's no such thing as series anymore (if they get the error)? 

```suggestion
		return "", errors.Errorf("unable to determine series for %q", arg.ApplicationName)
```

Reply:
Done